fix decode_png function.

yoya opened the pull request:
Problem: In the case of palette png & args channels=0, force into grayscale.
https://github.com/tensorflow/tensorflow/issues/20028

yoya committed:
fix decode_png function.
Problem: In the case of palette png & channels=0, force into grayscale.

caisq commented:
@yoya Thanks for sending this PR. Can you please sign the CLA?

tensorflowbutler commented:
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @caisq: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

skye commented on tensorflow/core/lib/png/png_io.cc:
```diff
@@ -232,11 +232,19 @@ bool CommonInitDecode(StringPiece png_string, int desired_channels,
   const bool has_tRNS =
       (png_get_valid(context->png_ptr, context->info_ptr, PNG_INFO_tRNS)) != 0;
+  if (context->channels == 0) {  // Autodetect number of channels
+    if (context->color_type == PNG_COLOR_TYPE_PALETTE) {
```
I'm not familiar with this code -- why do we need to handle PNG_COLOR_TYPE_PALETTE specially? Is it because we convert palette to rgb below?

yoya replied:
Certainly, you convert palette to rgb bellow.
But grayscale conversion run after that, because channels is 1 (args channels=0)

This issue depends on meaning of:
 "the number of channels in the PNG-encoded image."
https://www.tensorflow.org/api_docs/python/tf/image/decode_png

png_get_channels function return 1 if colortype is PALETTE.
http://www.libpng.org/pub/png/libpng-manual.html#section-3.6
Palette has indices channels so channels 1 only.

I think that the current behavior is unkind.

skye replied:
Ah I see, thank you for the explanation. I hadn't looked at the documentation before, but now it makes sense.

skye reviewed (changes requested):
Can you add some test coverage?

I suggest adding a (small) palette png here: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/lib/png/testdata

And then adding it to this test case: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py#L3204

(You'll also have to add the image file here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/BUILD#L4540)

Lemme know if this is difficult for some reason.

yoya commented:
Thank you for telling me how to add tests.
Please wait for a while.

I'll add the follwing two image files & test settings.
- lena_palette.png
- lena_palette_trns.png

skye commented:
Sounds good, thanks. Please let me know when you're done.

yoya committed:
- adding PNG-8(palette) files.
convert lena_std.png -scale '51x26!' png8:lena_palette.png
convert lena_rgba.png png8:lena_palette_trns.png

- update python test script
- update core/BUILD file

yoya commented:
https://github.com/tensorflow/tensorflow/pull/20029/commits/b28b82052525517beff238d2c53bde0171a4a59f
I added two PNG palette image files to the test coverage.

skye commented on tensorflow/python/ops/image_ops_test.py:
```diff
@@ -3105,7 +3105,7 @@ class PngTest(test_util.TensorFlowTestCase):
     # Read some real PNGs, converting to different channel numbers
     prefix = "tensorflow/core/lib/png/testdata/"
-    inputs = (1, "lena_gray.png"), (4, "lena_rgba.png")
+    inputs = (1, "lena_gray.png"), (4, "lena_rgba.png"), (3, "lena_palette.png"), (4, "lena_palette_trns.png")
```
This line is too long

skye reviewed (dismissed):
Looks good except for the long line, thanks!

yoya committed:
fixing to coding style, too long line.

yoya commented:
https://github.com/tensorflow/tensorflow/pull/20029/commits/1f31116a081afbf6794c5247a0b3b8ac120896b0
Excuse me. I added a line break.

skye reviewed (dismissed):
Ah sorry, didn't realize I had to re-approve.

skye commented:
Jenkins test this please (I think this should work...)

skye commented:
@caisq can you help me get this committed? I'm not familiar with the PR process

yoya committed:
$ mogrify +matte tensorflow/core/lib/png/testdata/lena_palette.png
(remove tRNS chunk)

skye commented:
Jenkins test this please
